app/bundles/CoreBundle/Form/Validator/Constraints/CircularDependencyValidator.php

```diff
@@ -2,72 +2,79 @@
 {
-    public function __construct(
-        private ListModel $model,
-        private RequestStack $requestStack
-    ) {
-    }
-
-    /**
-     * @param array $filters
-     */
-    public function validate($filters, Constraint $constraint): void
+    public function __construct(private SegmentDependencyTreeFactory $segmentDependencyTreeFactory)
     {
-        $dependentSegmentIds = $this->flatten(array_map(fn ($id) => $this->reduceToSegmentIds($this->model->getEntity($id)->getFilters()), $this->reduceToSegmentIds($filters)));
-
-        try {
-            $segmentId = $this->getSegmentIdFromRequest();
-            if (in_array($segmentId, $dependentSegmentIds)) {
-                $this->context->addViolation($constraint->message);
-            }
-        } catch (\UnexpectedValueException) {
-            // Segment ID is not in the request. May be new segment.
-        }
+        $this->segmentDependencyTreeFactory = $segmentDependencyTreeFactory;
     }
 
     /**
-     * @throws \UnexpectedValueException
+     * @param LeadList|mixed $segment
      */
-    private function getSegmentIdFromRequest(): int
+    public function validate($segment, Constraint $constraint)
     {
-        $request     = $this->requestStack->getCurrentRequest();
-        $routeParams = $request->get('_route_params');
+        if (!$constraint instanceof CircularDependency) {
+            throw new UnexpectedTypeException($constraint, CircularDependency::class);
+        }
 
-        if (empty($routeParams['objectId'])) {
-            throw new \UnexpectedValueException('Segment ID is missing in the request');
+        if (!$segment instanceof LeadList) {
+            return;
         }
 
-        return (int) $routeParams['objectId'];
-    }
+        $parentNode     = $this->segmentDependencyTreeFactory->buildTree($segment);
+        $directChildren = $parentNode->getChildrenArray();
 
-    private function reduceToSegmentIds(array $filters): array
-    {
-        $segmentFilters = array_filter($filters, fn (array $filter): bool => 'leadlist' === $filter['type']
-            && in_array($filter['operator'], [OperatorOptions::IN, OperatorOptions::NOT_IN]));
+        $segmentIdsToCheck = array_map(fn ($child) => $child->getValue(), $directChildren);
 
-        $segentIdsInFilter = array_map(function (array $filter) {
-            $bcValue = $filter['filter'] ?? [];
+        foreach ($directChildren as $directChildNode) {
+            /** @var \RecursiveIteratorIterator<IntNode> $iterator */
+            $iterator = new \RecursiveIteratorIterator($directChildNode, \RecursiveIteratorIterator::LEAVES_ONLY);
+            foreach ($iterator as $childNode) {
+                if (in_array($childNode->getValue(), $segmentIdsToCheck)) {
+                    $this->context->buildViolation($constraint->message)
+                        ->atPath('filters')
+                        ->setCode((string) Response::HTTP_UNPROCESSABLE_ENTITY)
```

Comment:
We should use a violation code instead of HTTP in the validators

Reply:
Which code would you suggest?

Reply:
I recommend defining a new `CircularDependency` constraint that includes a unique error code and a specific error message for this validator. You can check the existing `NotWeakValidator` to see how it's done. Handling the HTTP response code should ideally be done within a controller if it's possible.

Reply:
I've missed that `CircularDependency` constraint already exists, so we just need an unique code and mapped message within `$errorNames`.